### DISCUSSION TITLE
Static site builders

### DISCUSSION THREAD

**Discussion by aehlke** [live-codes/livecodes#573] [Ideas]
Hi again

I've discovered Zola as an attractive static site builder alternative to Hugo, which can't really run in-browser. Zola however is Rust and someone ported it to WASM to build sites in-browser a year ago: https://dstaley.com/posts/running-zola-on-wasm/

Would be very cool to have this working in live-codes!

Another interesting one is Wordpress WASM: https://github.com/WordPress/wordpress-playground which has recently done work to enable simple ways of generating a static site

**Comment by hatemhosny** [comment]
Hi @aehlke 
Nice to have more suggestions and ideas from you :)

How would you see these integrated in LiveCodes?
How would that be different from the export functionality that is already available?

**Reply by aehlke** [reply]
I don't know how they expect file trees to be inputted when running in WASM. LiveCodes would be able to have the project installed and running nicely which is not packaged in any up to date way besides the rought cut from that blog post.

WP WASM is GPL however. Zola would be most interesting. I haven't seen a live in-browser demo of running Zola, you always have to install something to try it which is high friction.

Personally I'd use this like the other languages you support to let users try Zola (or other similar ideas - I couldn't find anything for other popular ones like Hugo, Jekyll, etc which all require a server environment it seems) and then I'd provide tools for easily deploying that to staging / prod envs

These would need to be automated also so either via a script automating livecodes instances, or extending the script inside livecodes. I wouldn't use the export GUIs and ideally would hide those things and give native experiences for that, and script the instance.

I see something like Zola (or WP, which you may not need to wrap) more interesting than Astro, Svelte because as WASM it's able to generate a whole site from a template and plugins, not only demoing a playground version of itself that can only act on one file at a time. Pretty powerful for WASM and the only two I've found so far really.

Thank you

**Reply by aehlke** [reply]
(My use case is to wrap deployable static site generators in native iOS and later Android apps, such that simple native editing UI can preview the results on-device and offline, and realtime sync to a prod environment, in addition to being deployed to CI as-is and not only running on device)

**Reply by hatemhosny** [reply]
I'm still not sure how this fits in LiveCodes. 
Static site generators converts inputs files (markdown, templates, etc) to multiple static files (e.g. html, css)
LiveCodes already converts markdown and multiple template formats to html. This can be exported.
However, there is no concept of file system (with multiple files) in LiveCodes. 
I do not think Zola fits as one of the supported language compilers, or can fit as a supported feature (like import, export, embed, etc)

**Reply by aehlke** [reply]
@hatemhosny after thinking it over I don't have a good use case for you, thanks for considering it. I started from the point of view that your tool is a great unified interface sandbox for executing wasm developer tools where there are content/style/logic components and hadn't fully connected the dots to how these static site wasm tools could fit in alongside ones like Svelte and Astro, given they both have input and rendered output. If Zola in particular had a good use case for you, then it'd be great to have it working inside a reliable sandbox, but I agree it doesn't make sense for your project

Maybe there are one-page use cases for WordPress or Zola where you have single file friendly inputs and outputs, and which can be used in parallel with other livecodes instances for rendering a handful of things or varieties of inputs to an instance opened to a particular script. That might fit your project but would require more research to develop the use cases. Zola lacks an online demo sandbox like LiveCodes, and is another wasm tool that can work on single file inputs and outputs.

**Reply by hatemhosny** [reply]
Thank you @aehlke 
Please keep coming up with bright ideas and suggestions. 
I always find them useful. 
Thank you 💚

**Reply by aehlke** [reply]
I will keep it coming, thanks. I'm using this as a user extensible plugin system besides powering other dynamic code execution features on iOS/macOS and web/android soon, in all my products. It's great. And now Apple recently made a clear new rule that user script plugins are fine in all apps, so your tool can be embedded into social apps in the App Store. You might want to release a vanilla wrapper around your app in fact, or I could sometime
